File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix) - 1
        cols = len(matrix[0]) - 1
        l, r = 0, rows
        while l <= r : 
            mid = l + ((r -l) // 2)
            if target < matrix[mid][0]:
                r = mid - 1
            elif target > matrix[mid][-1]:
                l = mid + 1
            else:
                break
        row = l + ((r - l)//2)
        l, r = 0, cols
        while l <= r:
            mid = l + ((r - l)//2)
            if target < matrix[row][mid]:
                r = mid - 1
            elif target > matrix[row][mid]:
                l = mid + 1
            else:
                return True
        return False
        

File: Data Structures & Algorithms/islands-and-treasure/submission-4.java
class Solution {
    public void islandsAndTreasure(int[][] grid) {
        int[][] directions = {{1,0}, {0,1}, {-1,0}, {0,-1}};
        int rows = grid.length, cols = grid[0].length,dist = 0;
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[rows][cols];
        for(int r = 0; r < rows;r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == 0){
                    q.add(new int[]{r,c});
                    visited[r][c] = true;
                }
            }
        }
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                int[] temp = q.poll();
                int row = temp[0], col = temp[1];
                grid[row][col] = dist;
                for(int[] dir : directions){
                    addToQueue(grid, row + dir[0], col + dir[1], q, visited);
                }
            }
            dist++;
        }
    }
    private void addToQueue(int[][] grid, int r, int c, Queue<int[]> q, boolean[][] visited){
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] == -1 || visited[r][c]){
            return;
        }
        q.add(new int[]{r,c});
        visited[r][c] = true;
    }
}
